File: app/apple/ios/HiAgentIOS/AppModel.swift
import Combine
import Foundation
import UIKit
import UniformTypeIdentifiers

@MainActor
final class AppModel: ObservableObject {
    /// The one roster in the process. A `@StateObject` alone was enough while the
    /// only way into the app was its own window; [`ShowScreenIntent`] runs in this
    /// process too but outside SwiftUI, with no environment to be handed, and it
    /// must reach the *same* roster the screen is showing.
    static let shared = AppModel()

    @Published private(set) var entries: [RosterEntry] = []
    @Published var selectedID: String?
    @Published private(set) var isRefreshing = false
    @Published var addRequest: AddAgentRequest?
    @Published var addLinkError: String?
    @Published private(set) var credentialRevision = 0
    /// Where the last thing the person handed over got to, or `nil` if they have not
    /// handed anything over this launch. The bytes behind a failure are not here —
    /// they are in the queue on disk ([`HandedDrop`]), which is what makes "Try
    /// again" survive the app being killed.
    @Published private(set) var handoffState: HandoffState?

    private let defaults = UserDefaults.standard
    private let keychain = KeychainStore()
    private let storageKey = "hi.agent.ios.roster.v1"

    init() {
        load()
    }

    func entry(id: String) -> RosterEntry? {
        entries.first(where: { $0.id == id })
    }

    func handleIncomingURL(_ url: URL) {
        do {
            addRequest = try AddAgentRequest(url: url)
            addLinkError = nil
        } catch {
            addLinkError = error.localizedDescription
        }
    }

    /// Ask an agent, by name, to let this device in.
    ///
    /// Nothing is stored by this call: what comes back is a code to show and a
    /// secret to wait on. The roster only grows once [`join`] lands.
    func askToJoin(name: String) async throws -> JoinInvitation {
        let baseURL = try CoreClient.address(forName: name)
        let ticket = try await CoreClient.askToJoin(at: baseURL, label: UIDevice.current.name)
        return JoinInvitation(
            baseURL: baseURL,
            label: agentLabel(for: name, at: baseURL),
            code: ticket.code,
            secret: ticket.secret
        )
    }

    /// Wait out one invitation, and take the credential the moment it is approved.
    ///
    /// Polls until answered or cancelled — the caller's `Task` is what ends it, so
    /// closing the sheet stops the wait. Past the request's own ten-minute life the
    /// agent answers `expired` and this throws, which is the same clock rather than
    /// a second one kept here.
    func join(_ invitation: JoinInvitation) async throws {
        while true {
            try Task.checkCancellation()
            switch try await CoreClient.pollJoin(at: invitation.baseURL, secret: invitation.secret) {
            case .approved(let credential):
                // From here it is the ordinary add: the credential goes to the
                // keychain, the agent joins the roster, and the stage opens it.
                try await add(
                    baseURL: invitation.baseURL.absoluteString,
                    code: credential,
                    label: invitation.label
                )
                return
            case .denied:
                throw CoreClientError.requestFailed("That was turned down.")
            case .expired:
                throw CoreClientError.requestFailed("Nobody answered in time. Ask again.")
            case .waiting:
                try await Task.sleep(for: .seconds(2))
            }
        }
    }

    func add(baseURL rawBaseURL: String, code rawCode: String, label rawLabel: String) async throws {
        let baseURL = try CoreClient.normalizeBaseURL(rawBaseURL)
        let code = rawCode.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !code.isEmpty else {
            throw CoreClientError.requestFailed("Enter the one-time code the agent is showing.")
        }
        let requestedLabel = rawLabel.trimmingCharacters(in: .whitespacesAndNewlines)
        let coreLabel = requestedLabel.isEmpty ? defaultCoreLabel(for: baseURL) : requestedLabel

        let result = try await CoreClient.exchange(
            at: baseURL,
            presented: code,
            label: UIDevice.current.name
        )
        let credential = result.exchange.credential ?? code
        try keychain.save(credential, account: credentialAccount(id: result.exchange.id))
        credentialRevision += 1

        let wasAttached = entries.first(where: { $0.id == result.exchange.id })?.attached
            ?? entries.isEmpty
        let entry = RosterEntry(
            id: result.exchange.id,
            label: coreLabel,
            baseURL: baseURL.absoluteString,
            attached: wasAttached
        )
        entries.removeAll(where: { $0.id == entry.id })
        entries.append(entry)
        if entry.attached {
            selectedID = entry.id
        }
        persist()
        await refresh(entryID: entry.id)
    }

    func attach(_ id: String) {
        guard entries.contains(where: { $0.id == id }) else {
            return
        }
        entries = entries.map { entry in
            var updated = entry
            updated.attached = entry.id == id
            return updated
        }
        selectedID = id
        persist()
    }

    func forget(_ id: String) throws {
        let removedWasAttached = entries.first(where: { $0.id == id })?.attached == true
        try keychain.delete(account: credentialAccount(id: id))
        entries.removeAll(where: { $0.id == id })
        if removedWasAttached, !entries.isEmpty {
            entries[0].attached = true
        }
        if selectedID == id || selectedID == nil {
            selectedID = entries.first(where: { $0.attached })?.id ?? entries.first?.id
        }
        persist()
    }

    func refresh() async {
        isRefreshing = true
        defer { isRefreshing = false }
        for entry in entries {
            await refresh(entryID: entry.id)
        }
    }

    func open(_ id: String) async throws -> CoreSession {
        guard let entry = entry(id: id),
              let baseURL = URL(string: entry.baseURL)
        else {
            throw CoreClientError.invalidAddress
        }
        let credential = try keychain.read(account: credentialAccount(id: id))
        let result = try await CoreClient.exchange(at: baseURL, presented: credential, label: entry.label)
        if let index = entries.firstIndex(where: { $0.id == id }) {
            entries[index].health = .here
        }
        return CoreSession(entryID: id, baseURL: baseURL, cookie: result.cookie)
    }

    func refresh(entryID: String) async {
        guard let index = entries.firstIndex(where: { $0.id == entryID }),
              let baseURL = URL(string: entries[index].baseURL)
        else {
            return
        }
        entries[index].health = .checking
        let state = await CoreClient.health(at: baseURL)
        guard let index = entries.firstIndex(where: { $0.id == entryID }) else {
            return
        }
        entries[index].health = state
    }

    // MARK: - Handing things over

    /// Hand a screenshot to the attached core (see [`ShowScreen`]).
    ///
    /// It goes onto the same queue a share does rather than straight out over the
    /// wire. The app is on screen either way, so the queue buys one thing here: a
    /// screen that failed to send is still on disk tomorrow. It used to live in a
    /// property and die with the process.
    func showScreen(data: Data, type: UTType?, note: String?) async {
        do {
            let drop = try HandedDrop.enqueue(
                note: note ?? ShowScreen.note,
                text: nil,
                files: [
                    HandedDrop.PendingFile(
                        filename: ShowScreen.filename(for: type, at: Date()),
                        mime: ShowScreen.mime(for: type),
                        source: .data(data)
                    )
                ]
            )
            // `nil` means no group container, which is a missing entitlement rather
            // than a runtime condition. Said out loud because the alternative is the
            // worst shape a failure can have here: the gesture appears to work, and
            // nothing ever arrives.
            guard drop != nil else {
                handoffState = .failed(reason: HandedDropError.noContainer.localizedDescription)
                return
            }
        } catch {
            handoffState = .failed(reason: error.localizedDescription)
            return
        }
        await deliverQueued()
    }

    /// Send everything waiting, oldest first.
    ///
    /// Called when the app comes forward, and on the `hiagent://shared` the share
    /// extension opens — **not on a timer and not from a background task.** The queue
    /// is drained where its result can be seen, because a failure that nobody is
    /// looking at is a file that quietly stops existing.
    ///
    /// Reports rather than throws: the caller is a scene phase or a URL, neither of
    /// which has anywhere to put an error. The banner does.
    func deliverQueued() async {
        let drops = HandedDrop.queued()
        guard !drops.isEmpty else {
            return
        }
        guard let entry = attachedEntry else {
            handoffState = .failed(
                reason: "This device hasn't been added to an agent yet, so there's nobody to hand it to."
            )
            return
        }
        guard let baseURL = URL(string: entry.baseURL) else {
            handoffState = .failed(reason: CoreClientError.invalidAddress.localizedDescription)
            return
        }

        let credential: String
        do {
            credential = try keychain.read(account: credentialAccount(id: entry.id))
        } catch let error as KeychainError {
            if case .read = error {
                handoffState = .failed(
                    reason: "This device no longer has access to \(entry.label). Add it again."
                )
            } else {
                handoffState = .failed(reason: error.localizedDescription)
            }
            return
        } catch {
            handoffState = .failed(reason: error.localizedDescription)
            return
        }

        handoffState = .sending(count: drops.reduce(0) { $0 + $1.itemCount })
        var sent = 0
        var failure: String?
        for drop in drops {
            do {
                // Words first. When a drop has both — a link shared with a picture of
                // it — the sentence lands ahead of the artifact, which is the order
                // the person did it in.
                if let text = drop.manifest.text, !text.isEmpty {
                    try await CoreClient.say(at: baseURL, credential: credential, text: text)
                    sent += 1
                }
                for part in drop.manifest.parts {
                    try await CoreClient.hand(
                        at: baseURL,
                        credential: credential,
                        body: drop.url(for: part),
                        boundary: drop.manifest.boundary
                    )
                    sent += 1
                }
                // Only now: a drop that died halfway is retried whole rather than
                // half-delivered and forgotten. The cost of that choice is a possible
                // duplicate; the cost of the other is a file the person believes they
                // sent.
                drop.discard()
            } catch {
                // **Kept, and the others are still tried.** Nothing here can tell a
                // refusal apart from a tunnel, and discarding on the wrong guess
                // throws away something a person chose to send. Carrying on past it
                // is the other half: a drop that can never land must not become a
                // wall that everything shared afterwards queues up behind.
                failure = failure ?? error.localizedDescription
            }
        }
        if let failure {
            handoffState = .failed(reason: failure)
            return
        }
        handoffState = .sent(coreLabel: entry.label, count: sent)
    }

    /// Try the queue again. The drops are still on disk, so this is a retry and not a
    /// request that the person go back and share it a second time.
    func retryHandoff() async {
        await deliverQueued()
    }

    func dismissHandoff() {
        handoffState = nil
    }

    /// The core everything this device does goes to — the same one the stage shows.
    var attachedEntry: RosterEntry? {
        if let selectedID, let entry = entry(id: selectedID) {
            return entry
        }
        return entries.first(where: { $0.attached }) ?? entries.first
    }

    private func load() {
        guard let data = defaults.data(forKey: storageKey) else {
            return
        }
        do {
            entries = try JSONDecoder().decode([RosterEntry].self, from: data)
            selectedID = entries.first(where: { $0.attached })?.id
        } catch {
            defaults.removeObject(forKey: storageKey)
        }
    }

    private func persist() {
        do {
            defaults.set(try JSONEncoder().encode(entries), forKey: storageKey)
        } catch {
            // The roster contains only small local metadata. A failed write should
            // not discard the in-memory selection or the credential in Keychain.
        }
    }

    private func credentialAccount(id: String) -> String {
        "credential.\(id)"
    }

    private func defaultCoreLabel(for baseURL: URL) -> String {
        let host = baseURL.host ?? "Agent"
        let path = baseURL.path.trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        return path.isEmpty ? host : "\(host)/\(path)"
    }

    /// What to call an agent added by name. The name itself when that is what was
    /// typed — "iloahz" reads better in the roster than "iloahz.hi-agent.xyz" — and
    /// the host when somebody pasted a whole address.
    private func agentLabel(for name: String, at baseURL: URL) -> String {
        let typed = name.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        let bare = typed.hasPrefix("@") ? String(typed.dropFirst()) : typed
        if !bare.isEmpty, !bare.contains("."), !bare.contains("/") {
            return bare
        }
        return defaultCoreLabel(for: baseURL)
    }
}

/// One outstanding "let me in", held only while the sheet is open.
struct JoinInvitation: Equatable {
    let baseURL: URL
    let label: String
    /// Shown here and on the agent's Reach view, for the two to be compared. It
    /// authorizes nothing.
    let code: String
    let secret: String
}

File: app/android/app/src/main/java/com/xiaoyuanzhu/hiagent/android/AppModel.kt
package com.xiaoyuanzhu.hiagent.android

import android.app.Application
import android.net.Uri
import android.os.Build
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.xiaoyuanzhu.hiagent.android.core.CoreClient
import com.xiaoyuanzhu.hiagent.android.core.CoreClientException
import com.xiaoyuanzhu.hiagent.android.core.CoreSession
import com.xiaoyuanzhu.hiagent.android.core.CredentialStore
import com.xiaoyuanzhu.hiagent.android.core.HandedFile
import com.xiaoyuanzhu.hiagent.android.core.HandoffState
import com.xiaoyuanzhu.hiagent.android.core.HealthState
import com.xiaoyuanzhu.hiagent.android.core.JoinState
import com.xiaoyuanzhu.hiagent.android.core.NetworkMonitor
import com.xiaoyuanzhu.hiagent.android.core.AddAgentLinkException
import com.xiaoyuanzhu.hiagent.android.core.AddAgentRequest
import com.xiaoyuanzhu.hiagent.android.core.RosterEntry
import com.xiaoyuanzhu.hiagent.android.core.RosterStore
import java.time.Instant
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import okhttp3.HttpUrl.Companion.toHttpUrlOrNull

/**
 * All client state the shell needs: which cores this device knows, which one is
 * attached, and how to open one. The core owns identity, credential issuance and
 * revocation, memory, cognition, and channel behaviour — none of that is here.
 */
class AppModel(application: Application) : AndroidViewModel(application) {
    private val roster = RosterStore(application)
    private val credentials = CredentialStore(application)
    private val network = NetworkMonitor(application)

    private val _entries = MutableStateFlow<List<RosterEntry>>(emptyList())
    val entries: StateFlow<List<RosterEntry>> = _entries.asStateFlow()

    private val _selectedId = MutableStateFlow<String?>(null)
    val selectedId: StateFlow<String?> = _selectedId.asStateFlow()

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing.asStateFlow()

    private val _addRequest = MutableStateFlow<AddAgentRequest?>(null)
    val addRequest: StateFlow<AddAgentRequest?> = _addRequest.asStateFlow()

    private val _addLinkError = MutableStateFlow<String?>(null)
    val addLinkError: StateFlow<String?> = _addLinkError.asStateFlow()

    /**
     * Bumped whenever a credential is written. The stage restarts its open on a
     * change, so adding an agent that was failing recovers without the person
     * having to find a Reload.
     */
    private val _credentialRevision = MutableStateFlow(0)
    val credentialRevision: StateFlow<Int> = _credentialRevision.asStateFlow()

    /**
     * Where the last thing handed over got to, or null if nothing has been this
     * launch — the banner is then absent, not empty. A share arrives while the face
     * is still painting, so without this the successful case and the "not paired with
     * anything" case look identical: the app opened, and nothing visibly happened.
     */
    private val _handoff = MutableStateFlow<HandoffState?>(null)
    val handoff: StateFlow<HandoffState?> = _handoff.asStateFlow()

    val isConnected: StateFlow<Boolean> = network.isConnected
        .stateIn(viewModelScope, SharingStarted.WhileSubscribed(5_000), true)

    init {
        val loaded = roster.load()
        _entries.value = loaded
        _selectedId.value = loaded.firstOrNull { it.attached }?.id
    }

    fun entry(id: String): RosterEntry? = _entries.value.firstOrNull { it.id == id }

    val current: RosterEntry?
        get() = _selectedId.value?.let { entry(it) }
            ?: _entries.value.firstOrNull { it.attached }
            ?: _entries.value.firstOrNull()

    // MARK: Adding an agent

    fun requestAdd(request: AddAgentRequest?) {
        _addRequest.value = request
    }

    fun clearAddLinkError() {
        _addLinkError.value = null
    }

    fun handleIncomingUri(uri: Uri) {
        try {
            _addRequest.value = AddAgentRequest.fromUri(uri)
            _addLinkError.value = null
        } catch (e: AddAgentLinkException) {
            _addLinkError.value = e.message
        }
    }

    /**
     * Ask an agent, by name, to let this device in.
     *
     * Nothing is stored by this call: what comes back is a code to show and a
     * secret to wait on. The roster only grows once [join] lands.
     */
    suspend fun askToJoin(name: String): JoinInvitation {
        val baseUrl = CoreClient.addressForName(name)
        val ticket = CoreClient.askToJoin(baseUrl, deviceName())
        return JoinInvitation(
            baseUrl = baseUrl.toString(),
            label = agentLabel(name, baseUrl.toString()),
            code = ticket.code,
            secret = ticket.secret,
        )
    }

    /**
     * Wait out one invitation, and take the credential the moment it is approved.
     *
     * Polls until answered or cancelled — the caller's coroutine is what ends it,
     * so leaving the screen stops the wait. Past the request's own ten-minute life
     * the agent answers `expired` and this throws, which is the same clock rather
     * than a second one kept here.
     */
    suspend fun join(invitation: JoinInvitation) {
        val baseUrl = CoreClient.normalizeBaseUrl(invitation.baseUrl)
        while (true) {
            when (val state = CoreClient.pollJoin(baseUrl, invitation.secret)) {
                is JoinState.Approved -> {
                    // From here it is the ordinary add: the credential goes to the
                    // Keystore, the agent joins the roster, and the stage opens it.
                    add(invitation.baseUrl, state.credential, invitation.label)
                    return
                }
                JoinState.Denied ->
                    throw CoreClientException.RequestFailed("That was turned down.")
                JoinState.Expired ->
                    throw CoreClientException.RequestFailed(
                        "Nobody answered in time. Ask again.",
                    )
                JoinState.Waiting -> delay(2_000)
            }
        }
    }

    /**
     * Exchange a one-time code — or a credential a [join] just claimed — and
     * remember the agent.
     *
     * `label` on the wire is what the *agent* will call this device; the roster
     * label is what this device calls the agent. They are different strings for
     * different readers, which is why only one of them is sent.
     */
    suspend fun add(rawBaseUrl: String, rawCode: String, rawLabel: String) {
        val baseUrl = CoreClient.normalizeBaseUrl(rawBaseUrl)
        val code = rawCode.trim()
        if (code.isEmpty()) {
            throw CoreClientException.RequestFailed(
                "Enter the one-time code the agent is showing.",
            )
        }
        val coreLabel = rawLabel.trim().ifEmpty { defaultCoreLabel(baseUrl.toString()) }

        val (exchange, _) = CoreClient.exchange(
            baseUrl = baseUrl,
            presented = code,
            label = deviceName(),
        )
        // Null means "keep what you have" — the core says that when a credential,
        // not a pairing code, was presented.
        val credential = exchange.credential ?: code
        credentials.save(credential, CredentialStore.account(exchange.id))
        _credentialRevision.value += 1

        val wasAttached = _entries.value.firstOrNull { it.id == exchange.id }?.attached
            ?: _entries.value.isEmpty()
        val entry = RosterEntry(
            id = exchange.id,
            label = coreLabel,
            baseUrl = baseUrl.toString(),
            addedAt = Instant.now().toString(),
            attached = wasAttached,
        )
        _entries.value = _entries.value.filterNot { it.id == entry.id } + entry
        if (entry.attached) {
            _selectedId.value = entry.id
        }
        persist()
        refresh(entry.id)
    }

    // MARK: Roster

    fun attach(id: String) {
        if (_entries.value.none { it.id == id }) return
        _entries.value = _entries.value.map { it.copy(attached = it.id == id) }
        _selectedId.value = id
        persist()
    }

    fun forget(id: String) {
        val removedWasAttached = _entries.value.firstOrNull { it.id == id }?.attached == true
        credentials.delete(CredentialStore.account(id))
        var remaining = _entries.value.filterNot { it.id == id }
        if (removedWasAttached && remaining.isNotEmpty()) {
            remaining = remaining.mapIndexed { index, entry ->
                entry.copy(attached = index == 0)
            }
        }
        _entries.value = remaining
        if (_selectedId.value == id || _selectedId.value == null) {
            _selectedId.value = remaining.firstOrNull { it.attached }?.id
                ?: remaining.firstOrNull()?.id
        }
        persist()
    }

    fun refreshAll() {
        viewModelScope.launch {
            _isRefreshing.value = true
            try {
                _entries.value.forEach { refresh(it.id) }
            } finally {
                _isRefreshing.value = false
            }
        }
    }

    suspend fun refresh(entryId: String) {
        val baseUrl = entry(entryId)?.baseUrl?.toHttpUrlOrNull() ?: return
        setHealth(entryId, HealthState.CHECKING)
        setHealth(entryId, CoreClient.health(baseUrl))
    }

    // MARK: Handing things over

    /**
     * Send what the person just shared to the attached core.
     *
     * **Straight out over the wire, with no queue** — unlike iOS, where a share
     * extension is a separate short-lived process that cannot reach the credential or
     * outlive its sheet. `ACTION_SEND` starts this activity, so by the time this runs
     * the app is open, the roster is loaded, the keychain is readable and there is a
     * conversation on screen to report into. There is nothing to park.
     *
     * Words go first when there are both: a link shared with a picture of it lands as
     * the sentence and then the artifact, which is the order the person did it in.
     *
     * Reports rather than throws: the caller is an intent, which has nowhere to put an
     * error.
     *
     * **A send that outlives the screen is not handled, and it is the known gap here.**
     * The upload runs in `viewModelScope` and reads through a `content://` grant that
     * belongs to the activity, so walking away from a large video mid-upload can lose
     * both: the scope is cancelled with the activity, and the grant is revoked with
     * the task. Small things — the photos and links that are almost all of sharing —
     * land long before that matters. Closing it means a foreground service holding the
     * upload and a copy of the bytes taken while the grant is live, which is the
     * Android shape of what iOS does with its on-disk queue.
     */
    fun share(text: String?, files: List<HandedFile>) {
        if (text.isNullOrBlank() && files.isEmpty()) return

        val entry = current
        if (entry == null) {
            _handoff.value = HandoffState.Failed(
                "This device hasn't been added to an agent yet, so there's nobody to hand it to.",
            )
            return
        }
        val baseUrl = entry.baseUrl.toHttpUrlOrNull()
        if (baseUrl == null) {
            _handoff.value = HandoffState.Failed("This agent's address is no longer usable.")
            return
        }

        val count = files.size + if (text.isNullOrBlank()) 0 else 1
        _handoff.value = HandoffState.Sending(count)
        viewModelScope.launch {
            try {
                val credential = credentials.read(CredentialStore.account(entry.id))
                if (!text.isNullOrBlank()) {
                    CoreClient.say(baseUrl, credential, text.trim())
                }
                if (files.isNotEmpty()) {
                    CoreClient.hand(baseUrl, credential, files)
                }
                _handoff.value = HandoffState.Sent(entry.label, count)
            } catch (e: Exception) {
                _handoff.value = HandoffState.Failed(
                    e.message ?: "That could not be sent.",
                )
            }
        }
    }

    fun dismissHandoff() {
        _handoff.value = null
    }

    /**
     * Open a session against one core. The stored credential goes to
     * `POST /api/session` and only the exchanged cookie comes back out — the
     * credential itself never leaves this class.
     */
    suspend fun open(id: String): CoreSession {
        val entry = entry(id) ?: throw CoreClientException.InvalidAddress(
            "This agent is no longer in the roster.",
        )
        val baseUrl = entry.baseUrl.toHttpUrlOrNull()
            ?: throw CoreClientException.InvalidAddress(
                "This agent's address is no longer usable.",
            )
        val credential = credentials.read(CredentialStore.account(id))
        val (_, cookie) = CoreClient.exchange(baseUrl, credential, entry.label)
        setHealth(id, HealthState.HERE)
        return CoreSession(entryId = id, baseUrl = baseUrl, cookie = cookie)
    }

    private fun setHealth(entryId: String, state: HealthState) {
        _entries.value = _entries.value.map {
            if (it.id == entryId) it.copy(health = state) else it
        }
    }

    private fun persist() = roster.save(_entries.value)

    private fun defaultCoreLabel(baseUrl: String): String {
        val url = baseUrl.toHttpUrlOrNull() ?: return "Agent"
        val path = url.encodedPath.trim('/')
        return if (path.isEmpty()) url.host else "${url.host}/$path"
    }

    /**
     * What to call an agent added by name. The name itself when that is what was
     * typed — "iloahz" reads better in the roster than "iloahz.hi-agent.xyz" — and
     * the host when somebody pasted a whole address.
     */
    private fun agentLabel(name: String, baseUrl: String): String {
        val bare = name.trim().lowercase().removePrefix("@")
        return if (bare.isNotEmpty() && !bare.contains('.') && !bare.contains('/')) {
            bare
        } else {
            defaultCoreLabel(baseUrl)
        }
    }

    /**
     * What the core will list this device as. Android has no
     * `UIDevice.current.name`, and since Android 10 the user-set device name is
     * not readable without permissions the app has no other use for — so this is
     * the marketing name, which is what a person recognises in a device list
     * anyway.
     */
    private fun deviceName(): String {
        val manufacturer = Build.MANUFACTURER.orEmpty().replaceFirstChar { it.uppercase() }
        val model = Build.MODEL.orEmpty()
        return when {
            model.isEmpty() -> manufacturer.ifEmpty { "Android device" }
            model.startsWith(manufacturer, ignoreCase = true) -> model
            manufacturer.isEmpty() -> model
            else -> "$manufacturer $model"
        }
    }
}

/** One outstanding "let me in", held only while the add screen is open. */
data class JoinInvitation(
    val baseUrl: String,
    val label: String,
    /**
     * Shown here and on the agent's Reach view, for the two to be compared. It
     * authorizes nothing.
     */
    val code: String,
    val secret: String,
)
